### PR TITLE
Allow mods to Override Rolling for negative traits

### PR DESCRIPTION
**Pull request by Grobobobo**

### CODE REVIEW

**Comment by Grobobobo** [comment]
Fixes #1150 .

**Comment by Grobobobo** [comment]
Should be fine now

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 1]
```diff
@@ -904,6 +904,7 @@ function bool RollForNegativeTrait(XComGameState NewGameState)
 	local int RollValue, WillPercentMark, HealthPercentMark;
 	local array<name> ValidTraits, GenericTraits;
 	local name TraitName;
+	local bool ShouldAcquireTrait;
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:907]:
```suggestion
	local bool ShouldAcquireTrait; // Variable for issue #1150
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 2]
```diff
@@ -912,8 +913,11 @@ function bool RollForNegativeTrait(XComGameState NewGameState)
 	HealthPercentMark = (100 - int((GetCurrentStat(eStat_HP) / GetMaxStat(eStat_HP)) * 100.0f));
 	ValidTraits.Length = 0;
 	
+	//Issue #1150
+	ShouldAcquireTrait = CanAcquireTrait(false) && RollValue < (WillPercentMark + HealthPercentMark);
+	
 	// Roll to see if they should gain a trait
-	if(CanAcquireTrait(false) && RollValue < (WillPercentMark + HealthPercentMark))
+	if(OverrideNegativeTraitRoll(ShouldAcquireTrait, NewGameState))
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:920]: The current code is a bit confusing to read (one would expect `if (OverrideNegativeTraitRoll())` then skip vanilla logic), so I suggest the following refactor:

```uc
// Start Issue #1150
// Roll to see if they should gain a trait
ShouldAcquireTrait = CanAcquireTrait(false) && RollValue < (WillPercentMark + HealthPercentMark); // Vanilla logic
ShouldAcquireTrait = OverrideNegativeTraitRoll(ShouldAcquireTrait, NewGameState);

if (ShouldAcquireTrait)
// End Issue #1150
{
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 3]
```diff
@@ -15691,6 +15695,37 @@ private function SetOverKillUnitValue(int OverKillDamage)
...
+/// ```event
+/// EventID: OverrideNegativeTraitRoll,
+/// EventData: [inout bool ShouldRollNegativeTrait],
+/// EventSource: XComGameState_Unit (UnitState),
+/// NewGameState: yes
+/// ```
+//
+private function bool OverrideNegativeTraitRoll(
+	bool ShouldRollNegativeTrait,
+	XComGameState NewGameState)
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:15713]: Put this function right after `RollForNegativeTrait` (where it's called from)

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 4]
```diff
@@ -912,8 +913,13 @@ function bool RollForNegativeTrait(XComGameState NewGameState)
...
 	ValidTraits.Length = 0;
 	
+	// Start Issue #1150
 	// Roll to see if they should gain a trait
-	if(CanAcquireTrait(false) && RollValue < (WillPercentMark + HealthPercentMark))
+	ShouldAcquireTrait = CanAcquireTrait(false) && RollValue < (WillPercentMark + HealthPercentMark); // Vanilla logic
+	ShouldAcquireTrait = OverrideNegativeTraitRoll(ShouldAcquireTrait, NewGameState);
+	
+	// Roll to see if they should gain a trait
+	if(ShouldAcquireTrait)
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:922]: Missing `// End Issue #1150` (line after the `if`)

**Review by Xymanek — CHANGES_REQUESTED** [review]
Please fix the comment omission and then approved

**Comment by Iridar** [comment]
LGTM. I tested the code in-game and it works, although seems to be somewhat limited in scope, meaning that it won't allow mods to make soldiers roll for negative traits after any mission under arbitrary conditions. 

The unit must have `if(UnitState.GetMentalState() == eMentalState_Tired || UnitState.GetMentalState() == eMentalState_Shaken)` before the `XComGameStateContext_StrategyGameRule::UpdateSoldierWill()` will call `XCGS_Unit::RollForNegativeTrait()`, where the override event is triggered.

**Review by Iridar — APPROVED** [review]

**Pull request merged**